pkg/component/networking/nodelocaldns/nodelocaldns.go

```diff
@@ -611,3 +813,172 @@ func (n *nodeLocalDNS) getAddress(useIPv6Brackets bool) string {
+		},
+		Spec: appsv1.DaemonSetSpec{
+			Selector: &metav1.LabelSelector{
+				MatchLabels: map[string]string{
+					labelKeyCleanUp: labelValueCleanUp,
+				},
+			},
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{
+						labelKeyCleanUp: labelValueCleanUp,
+					},
+				},
+				Spec: corev1.PodSpec{
+					HostNetwork:       true,
+					RestartPolicy:     corev1.RestartPolicyAlways,
+					PriorityClassName: "system-node-critical",
+					Tolerations: []corev1.Toleration{
+						{
+							Operator: corev1.TolerationOpExists,
+							Effect:   corev1.TaintEffectNoExecute,
+						},
+						{
+							Operator: corev1.TolerationOpExists,
+							Effect:   corev1.TaintEffectNoSchedule,
+						},
+					},
+					NodeSelector: map[string]string{
+						labelKeyCleanUpRequired: "true",
+					},
+					Containers: []corev1.Container{
+						{
+							Name:  "cleanup",
+							Image: n.values.AlpineImage,
+							Command: []string{
+								"/bin/sh",
+								"-c",
+								"/scripts/cleanup.sh",
+							},
+							VolumeMounts: []corev1.VolumeMount{
+								{
+									Name:      volumeMountNameCleanUp,
+									MountPath: volumeMountPathCleanUp,
+									ReadOnly:  true,
+								},
+								{
+									Name:      volumeMountNameXtablesLock,
+									MountPath: volumeMountPathXtablesLock,
+									ReadOnly:  false,
+								},
+							},
+							ReadinessProbe: &corev1.Probe{
+								InitialDelaySeconds: 10,
+								PeriodSeconds:       5,
+								SuccessThreshold:    1,
+								FailureThreshold:    3,
+								ProbeHandler: corev1.ProbeHandler{
+									Exec: &corev1.ExecAction{
+										Command: []string{
+											"cat",
+											"/tmp/healthy",
+										},
+									},
+								},
+							},
+							SecurityContext: &corev1.SecurityContext{
+								Capabilities: &corev1.Capabilities{
+									Add: []corev1.Capability{"NET_ADMIN", "NET_RAW"},
+								},
+								Privileged: ptr.To(false),
+							},
+						},
+					},
+					Volumes: []corev1.Volume{
+						{
+							Name: volumeMountNameCleanUp,
+							VolumeSource: corev1.VolumeSource{
+								ConfigMap: &corev1.ConfigMapVolumeSource{
+									LocalObjectReference: corev1.LocalObjectReference{
+										Name: cleanupConfigMapName,
+									},
+									DefaultMode: ptr.To[int32](0775),
+								},
+							},
+						},
+						{
+							Name: volumeMountNameXtablesLock,
+							VolumeSource: corev1.VolumeSource{
+								HostPath: &corev1.HostPathVolumeSource{
+									Path: volumeMountPathXtablesLock,
+									Type: ptr.To(corev1.HostPathFileOrCreate),
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+// waitForDaemonSetCompletion waits until the DaemonSet has all desired pods ready.
+// It polls the DaemonSet status until the desired number of pods are ready and none are unavailable.
+func waitForDaemonSetCompletion(ctx context.Context, client client.Client, namespace, name string) error {
```

Comment:
Not sure if we should improve on the naming of the function. It makes sense in the `Destroy(...)` function, but the comment sounds a bit weird because it equals completion with ready, which is not generally true. However, `waitForDaemonSetReady(...)` would be somewhat less readable in `Destroy(...)`.
WDYT?